### PR TITLE
Add unit tests for `metrics::record`

### PR DESCRIPTION
**Pull request by hawkw**
This PR adds unit tests for `metrics::record`, based on the benchmarks for the
same function. Currently, there is a test that fires a single response end event
and asserts that the metrics state is correct afterward, and a test that fires
all the events to simulate a full connection lifetime, and asserts that the
metrics state is correct afterward. I'd like to also add a test that simulates
multiple events with different labels, but I'll add that in a subsequent PR,

In order to add these tests, it was necessary to make some fields in metrics
structs `pub` so that the test can access them. These fields should only be
exposed within the metrics module, however. I also added some test-only
functions to `metrics::Histogram`s, to make them easier to make assertions
about.

### CODE REVIEW

File: proxy/src/telemetry/metrics/transport.rs [hunk 1]
```diff
@@ -17,18 +17,18 @@ pub(super) type OpenScopes = Scopes<TransportLabels, Stamped<OpenMetrics>>;
...
 #[derive(Debug, Default)]
 pub(super) struct OpenMetrics {
-    open_total: Counter,
-    open_connections: Gauge,
-    write_bytes_total: Counter,
-    read_bytes_total: Counter,
+    pub open_total: Counter,
+    pub open_connections: Gauge,
+    pub write_bytes_total: Counter,
+    pub read_bytes_total: Counter,
```
Comment (olix0r) [inline proxy/src/telemetry/metrics/transport.rs]: it's unfortunate that we have to change the interface for the tests.

Wdyt about having a 

```
#[cfg(test)]
impl OpenMetrics {
   ...
}
```

with helpers?

Comment (hawkw) [reply proxy/src/telemetry/metrics/transport.rs]: Yeah, that sounds good. Will do.

File: proxy/src/telemetry/metrics/histogram.rs [hunk 2]
```diff
@@ -88,6 +88,113 @@ impl<V: Into<u64>> Histogram<V> {
...
         self.sum += value;
     }
+
+    // ===== Test-only methods to help with assertions about histograms. =====
+
+    /// Assert the bucket containing `le` has a count of at least `at_least`.
+    #[cfg(test)]
+    pub fn assert_bucket_at_least(&self, le: u64, at_least: u64) {
+        for (&bucket, &count) in self {
+            let is_le = if let Bucket::Le(ceiling) = bucket {
```
Comment (olix0r) [inline proxy/src/telemetry/metrics/histogram.rs]: Can you factor out the repeated `is_le` logic?  Tests would be more readable with `if le < bucket { ... }` or something that effect.

File: proxy/src/telemetry/metrics/http.rs [hunk 3]
```diff
@@ -16,15 +16,15 @@ pub(super) type RequestScopes = Scopes<RequestLabels, Stamped<RequestMetrics>>;
...
 }
 
 pub(super) type ResponseScopes = Scopes<ResponseLabels, Stamped<ResponseMetrics>>;
 
 #[derive(Debug, Default)]
 pub struct ResponseMetrics {
-    total: Counter,
-    latency: Histogram<latency::Ms>,
+    pub(super) total: Counter,
+    pub(super) latency: Histogram<latency::Ms>,
```
Comment (olix0r) [inline proxy/src/telemetry/metrics/http.rs]: similarly, it would be nice if we didn't have to expose these internals outside of tests

Comment (hawkw) [reply proxy/src/telemetry/metrics/http.rs]: Yeah, I'll add test-only accessors.

**Comment by hawkw** [comment]
@olix0r, efa51aa7a5639b4e217a2271ad35116dd7c9811c and c368124e2231502a6ec2c01f4dcb95df669501ec make the changes you've requested.

File: proxy/src/telemetry/metrics/histogram.rs [hunk 4]
```diff
@@ -88,6 +88,98 @@ impl<V: Into<u64>> Histogram<V> {
...
+        }
+        self
+    }
+
+    /// Assert all buckets less than the one containing `value` have
+    /// counts of exactly `exactly`.
+    #[cfg(test)]
+    pub fn assert_lt_exactly(&self, value: u64, exactly: u64) -> &Self {
+        for (i, &bucket) in self.bounds.0.iter().enumerate() {
+            if let Bucket::Le(ceiling) = bucket {
```
Comment (olix0r) [inline proxy/src/telemetry/metrics/histogram.rs]: tioli: i find the big block and then `} else { break ]` to be a bit tricky to folow flow-control-wise.  Consider:

```rust
let ceiling = match bucket {
  Bucket::Le(c) => c,
  Bucket::Inf => break,
};
```

all the info i need in once place!

**Pull request merged**